### PR TITLE
[MS Teams] add connector

### PR DESCRIPTION
**Pull request by EugeneLightsOn**
### What's being changed:
MS Teams connector 

### How did you test this change (include any code snippets, API requests, screenshots, or gifs):
- Tested script locally

### CODE REVIEW

File: msteams/README.md [hunk 1]
```diff
@@ -0,0 +1,162 @@
+# Microsoft Teams
+
+This package is a utility for connecting Cohere to Microsoft Teams.
+
+It uses Microsoft Graph API run the search query and return matching Teams chat messages.
```
Comment (tianjing-li) [inline msteams/README.md:5]: `to run the`

File: msteams/provider/client.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+
+import requests
+from flask import current_app as app, request
+from msal import ConfidentialClientApplication
+
+from . import UpstreamProviderError
+
+AUTHORIZATION_HEADER = "Authorization"
+BEARER_PREFIX = "Bearer "
+CACHE_SIZE = 256
```
Comment (tianjing-li) [inline msteams/provider/client.py:13]: Don't need `CACHE_SIZE` in this file

File: msteams/provider/client.py [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+                        },
+                    }
+                ]
+            },
+        )
+        if not response.ok:
+            raise UpstreamProviderError(
+                f"Error while searching Outlook: {response.text}"
+            )
+        for hit_container in response.json()["value"][0]["hitsContainers"]:
```
Comment (tianjing-li) [inline msteams/provider/client.py:166]: I think it's more clear to return the `response.json()["value"][0]["hitsContainers"]` here, and in the `provider.py` do the processing

File: msteams/provider/client.py [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+
+    async def _get_message(self, hit):
+        url = f"https://graph.microsoft.com/v1.0/chats/{hit['resource']['chatId']}/messages/{hit['resource']['id']}"
+        params = {
+            "$select": "id,subject,summary,body,from,createdDateTime,webUrl,attachments,eventDetail",
+        }
+
+        async with self.session.get(
+            url,
+            headers={"Authorization": f"Bearer {self.access_token}"},
```
Comment (tianjing-li) [inline msteams/provider/client.py:85]: In `set_user_access_token()` `self.headers` gets set, but in `set_app_access_token()` it doesn't, can we add that to the app auth and then everywhere the headers are needed you can use `self.headers` instead?

File: msteams/provider/client.py [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+
+    def _delegated_search(self, query):
+        results = []
+        response = requests.post(
+            self.SEARCH_URL,
+            headers={"Authorization": f"Bearer {self.access_token}"},
+            json={
+                "requests": [
+                    {
+                        "entityTypes": ["chatMessage"],
```
Comment (tianjing-li) [inline msteams/provider/client.py:153]: Can we add `["chatMessage"]` as a value in a class attribute? perhaps `SEARCH_ENTITY_TYPES`

File: msteams/provider/provider.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+        for result in results
+    ]
+
+
+def parse_results_attachments(results):
+    attachments_to_unstructured = []
+    for result in results:
+        attachments = result["attachments"]
+        if attachments:
+            for attachment in attachments:
```
Comment (tianjing-li) [inline msteams/provider/provider.py:47]: `attachments_to_unstructured.extend([attachment for attachment in attachments if attachment["contentType"] == "reference"])`

Comment (tianjing-li) [reply msteams/provider/provider.py:47]: In general, let's prevent heavily nested blocks of code

**Review by tianjing-li — CHANGES_REQUESTED** [review]
Thanks for working on this! Couple of comments

File: msteams/README.md [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+the appropriate permissions. The MSTEAMS_USER_ID represents the user ID of the individual who registered the app.
+To obtain the user ID, you can use the Microsoft Entra admin center Identity -> Users -> All Users menu and select
+your user. The user ID is the "Object ID" field. This information is essential for app-only authentication and is
+relevant to the limitations of the Microsoft Graph API search functionality.
+
+
+
+#### OAuth
+
+When using OAuth for authentication, the connector does not require any additional environment variables. Instead, the OAuth flow should occur outside of the Connector and Cohere's API will forward the user's access token to this connector through the `Authorization` header.
```
Comment (tianjing-li) [inline msteams/README.md:52]: I think it would require the `AUTH_TYPE` variable

File: msteams/README.md [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+
+With OAuth the connector will be able to search any Teams chat messages that the user has access to.
+
+To configure OAuth, follow the same steps in the Configuration section to create a Microsoft 365 App. You will also need to register a redirect URI on that app to `https://api.cohere.com/v1/connectors/oauth/token`.
+
+You can then register the connector with Cohere's API using the following configuration:
+Note: Your App key and App secret values correspond to `client_id` and `client_secret` respectively.
```
Comment (tianjing-li) [inline msteams/README.md:59]: I believe this was pasted from dropbox, MS teams doesn't have app key/secret, just the client_id/secret directly

File: msteams/README.md [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+  --header 'Authorization: Bearer {COHERE-API-KEY}' \
+  --header 'Content-Type: application/json' \
+  --data-raw '{
+  "name": "MS Teams with OAuth",
+  "url": "{YOUR_CONNECTOR-URL}",
+  "oauth": {
+    "client_id": "{Your App CLIENT-ID}",
+    "client_secret": "{Your App CLIENT-SECRET}",
+    "authorize_url": "https://login.microsoftonline.com/{Your APP Tenant ID}/oauth2/v2.0/authorize"
+    "token_url": "https://login.microsoftonline.com/{Your APP Tenant ID}/oauth2/v2.0/token"
```
Comment (tianjing-li) [inline msteams/README.md:74]: Nit: for consistency `Your App TENANT-ID`

File: msteams/README.md [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+    "scope": ".default offline_access"
+  }
+}'
+```
+
+
+To make a search requests to Microsoft Graph API, the connector needs to be authenticated with a token obtained using
+Oauth 2.0, and passed to the connector in the Authorization header with the Bearer schema.
```
Comment (tianjing-li) [inline msteams/README.md:82]: Can be open to interpretation, I would just mention that no more configuration is needed here and that after successfully registration, Cohere will take care of OAuth steps including passing in the correct headers

File: msteams/README.md (hunk 1 above)
Comment (EugeneLightsOn) [reply msteams/README.md]: Thanks

File: msteams/provider/client.py (hunk 2 above)
Comment (EugeneLightsOn) [reply msteams/provider/client.py]: Thank you.

File: msteams/provider/client.py (hunk 4 above)
Comment (EugeneLightsOn) [reply msteams/provider/client.py]: Good point - thanks. Done

File: msteams/provider/client.py (hunk 5 above)
Comment (EugeneLightsOn) [reply msteams/provider/client.py]: Done

File: msteams/provider/provider.py [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+                    attachments_to_unstructured.append(attachment)
+    if len(attachments_to_unstructured) > 0:
+        unstructured_client = get_unstructured_client()
+        unstructured_client.start_session()
+        unstructured_results = unstructured_client.batch_get(
+            attachments_to_unstructured
+        )
+        for result in results:
+            attachments = result["attachments"]
+            if attachments:
```
[the remark this answers is not in the document]
Comment (EugeneLightsOn) [reply msteams/provider/provider.py]: Thank you

File: msteams/README.md (hunk 7 above)
Comment (EugeneLightsOn) [reply msteams/README.md]: Thanks. Fixed

File: msteams/README.md (hunk 9 above)
Comment (EugeneLightsOn) [reply msteams/README.md]: Done

File: msteams/README.md [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+  --header 'Accept: */*' \
+  --header 'Authorization: Bearer {COHERE-API-KEY}' \
+  --header 'Content-Type: application/json' \
+  --data-raw '{
+  "name": "MS Teams with OAuth",
+  "url": "{YOUR_CONNECTOR-URL}",
+  "oauth": {
+    "client_id": "{Your App CLIENT-ID}",
+    "client_secret": "{Your App CLIENT-SECRET}",
+    "authorize_url": "https://login.microsoftonline.com/{Your App TENANT-ID}/oauth2/v2.0/authorize"
```
Comment (tianjing-li) [inline msteams/README.md:79]: Just missing some `,` after the authorize/token_urls

Comment (EugeneLightsOn) [reply msteams/README.md]: Thanks

**Review by tianjing-li — APPROVED** [review]
Approved

File: msteams/provider/unstructured.py [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+        self.get_content_url = f"{unstructured_base_url}/general/v0/general"
+        # API key optional if self-hosted
+        self.headers = {"unstructured-api-key": api_key} if api_key else {}
+        # Manually cache because functools.lru_cache does not support async methods
+        self.cache = OrderedDict()
+        self.start_session()
+
+    def start_session(self):
+        self.loop = asyncio.new_event_loop()
+        self.session = aiohttp.ClientSession(loop=self.loop)
```
[the remark this answers is not in the document]
Comment (EugeneLightsOn) [reply msteams/provider/unstructured.py]: Good point, Thank you Walter

File: msteams/pyproject.toml [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+[tool.poetry.dependencies]
+python = "^3.11"
+flask = "2.2.5"
+python-dotenv = "1.0.0"
+connexion = {extras = ["swagger-ui"], version = "^2.14.2"}
+msal = "^1.25.0"
+gunicorn = "^21.2.0"
+requests = "^2.31.0"
+beautifulsoup4 = "^4.12.2"
+unstructured-client = "^0.14.3"
```
[the remark this answers is not in the document]
Comment (EugeneLightsOn) [reply msteams/pyproject.toml:17]: Yes

File: msteams/provider/provider.py [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+
+
+def serialize_results(results):
+    return [
+        {
+            "id": str(result["id"]),
+            "title": str(
+                result["subject"]
+                or strip_html_tags(result["body"]["content"])[:30] + "..."
+            ),
```
[the remark this answers is not in the document]
Comment (EugeneLightsOn) [reply msteams/provider/provider.py]: Done

File: msteams/provider/client.py [hunk 16]
```diff
@@ -0,0 +1,246 @@
...
+        attachments = []
+        for message in messages:
+            if message is not None:
+                if "attachments" in message:
+                    for attachment in message["attachments"]:
+                        attachment["downloadUrl"] = None
+                        if attachment["contentType"] == "reference":
+                            if "contentUrl" in attachment:
+                                attachments.append(attachment)
+                results.append(message)
```
[the remark this answers is not in the document]
Comment (EugeneLightsOn) [reply msteams/provider/client.py]: Yeah, it looks better. Thank you, Walter. Done

**Pull request merged**